demos/quic/server/server.c

```diff
@@ -0,0 +1,230 @@
+#include <openssl/err.h>
+#include <openssl/ssl.h>
+#include <openssl/quic.h>
+#include <netinet/in.h>
+#include <unistd.h>
+#include <assert.h>
+
+/*
+ * This is a basic demo of QUIC server functionality in which one connection at
+ * a time is accepted in a blocking loop.
+ */
+
+/* ALPN string for TLS handshake */
+static const unsigned char alpn_ossltest[] = {
+    /* "\x08ossltest" (hex for EBCDIC resilience) */
+    0x08, 0x6f, 0x73, 0x73, 0x6c, 0x74, 0x65, 0x73, 0x74
+};
+
+/* This callback validates and negotiates the desired ALPN on the server side. */
+static int select_alpn(SSL *ssl,
+                       const unsigned char **out, unsigned char *out_len,
+                       const unsigned char *in, unsigned int in_len,
+                       void *arg)
+{
+    if (SSL_select_next_proto((unsigned char **)out, out_len,
+                              alpn_ossltest, sizeof(alpn_ossltest), in, in_len)
+            != OPENSSL_NPN_NEGOTIATED)
+        return SSL_TLSEXT_ERR_ALERT_FATAL;
+
+    return SSL_TLSEXT_ERR_OK;
+}
+
+/* Create SSL_CTX. */
+static SSL_CTX *create_ctx(const char *cert_path, const char *key_path)
+{
+    SSL_CTX *ctx;
+
+    ctx = SSL_CTX_new(OSSL_QUIC_server_method());
+    if (ctx == NULL)
+        goto err;
+
+    /* Load certificate and corresponding private key. */
+    if (SSL_CTX_use_certificate_chain_file(ctx, cert_path) <= 0) {
+        fprintf(stderr, "couldn't load certificate file: %s\n", cert_path);
```

Comment:
>SSL_CTX_use_certificate_file() loads the first certificate stored in file into ctx. The
>formatting type of the certificate must be specified from the known types SSL_FILETYPE_PEM,
       SSL_FILETYPE_ASN1.  SSL_use_certificate_file() loads the certificate from file into ssl.  See
       the NOTES section on why SSL_CTX_use_certificate_chain_file() should be preferred.

>SSL_CTX_use_certificate_chain_file() should be used instead of the
       SSL_CTX_use_certificate_file() function in order to allow the use of complete certificate
       chains even when no trusted CA storage is used or when the CA issuing the certificate shall not
       be added to the trusted CA storage.

Using `SSL_CTX_use_certificate_file` would offer no way to serve intermediate certificates... basically guaranteeing incorrect behaviour when using any certificate issued by a public CA.